Wrappers/Python/cil/optimisation/algorithms/PD3O.py

```diff
@@ -0,0 +1,145 @@
+        super(PD3O, self).__init__(**kwargs)
+
+              
+        self.set_up(f=f, g=g, h=h,  operator=operator, delta=delta, gamma=gamma, initial=initial, **kwargs)
+ 
+                  
+    def set_up(self, f, g, h, operator, delta=None, gamma=None, initial=None,**kwargs):
+        
+        logging.info("{} setting up".format(self.__class__.__name__, ))
+        
+        self.f = f # smooth function
+        if isinstance(self.f, ZeroFunction):
+            warnings.warn(" If self.f is the ZeroFunction, then PD3O = PDHG. Please use PDHG instead. Otherwise, select a relatively small parameter gamma ", UserWarning)
+            if gamma is None:
+                gamma = 1.0/operator.norm()                      
+        
+        self.g = g # proximable
+        self.h = h # composite
+        self.operator = operator
+        
+        if gamma is None:
+            gamma = 0.99*2.0/self.f.L
+        
+        if delta is None :
+            delta = 0.99/(gamma*self.operator.norm()**2)
+        
+        self.gamma = gamma
+        self.delta = delta  
+
+        if initial is None:
+            self.x = self.operator.domain_geometry().allocate(0)
+        else:
+            self.x = initial.copy()
+
+        self.x_bar = self.x.copy()    
+        self.x_old = self.operator.domain_geometry().allocate(0)
+        
+        self.s_old = self.operator.range_geometry().allocate(0)
+        self.s = self.operator.range_geometry().allocate(0)
+                
+        self.grad_f = self.operator.domain_geometry().allocate(0)        
+  
+        self.configured = True
+        logging.info("{} configured".format(self.__class__.__name__, ))
+        
+
+    def update(self):
+        r""" Performs a single iteration of the PD3O algorithm        
+        """
+
+        # Following equations 4 in https://link.springer.com/article/10.1007/s10915-018-0680-3
+        # in this case order of proximal steps we recover the (primal) PDHG, when f=0
+
+        
+        # proximal conjugate step
+        self.operator.direct(self.x_bar, out=self.s)
+        self.s_old.sapyb(1, self.s, self.delta, out=self.s_old)
+        self.h.proximal_conjugate(self.s_old, self.delta, out=self.s)
+        
+        # proximal step        
+        self.f.gradient(self.x_old, out=self.grad_f)
+        self.x_old.sapyb(1., self.grad_f, -self.gamma, out = self.grad_f) # x_old - gamma * grad_f(x_old)        
+        self.operator.adjoint(self.s, out=self.x_old)
+        self.x_old.sapyb(-self.gamma, self.grad_f, 1.0, out=self.x_old)
+        self.g.proximal(self.x_old, self.gamma, out = self.x)
+
+        # update step        
+        self.f.gradient(self.x, out=self.x_bar)                    
+        self.x_bar *= self.gamma
+        self.grad_f += self.x_bar
+        self.x.sapyb(2, self.grad_f, -1.0,  out=self.x_bar) # 2*x - x_old + gamma*(grad_f_x_old) - gamma*(grad_f_x)
+        
+        self.x_old.fill(self.x)      
+        self.s_old.fill(self.s)         

```

Comment:
I think I have done this

Reply:
I did not do it in an _update_previous_solution function because it might cause issues with using algo.x or algo.get_output() as we have overwritten x_old

Reply:
In PDHG we redefine it here.
https://github.com/TomographicImaging/CIL/blob/c87b61ea4b28ec8894ad7a665122eccd7fd6bde6/Wrappers/Python/cil/optimisation/algorithms/PDHG.py#L354C1-L356C26

But tbh I prefer it being more explicit in the update step. I think `_update_previous_solution` obfuscates things a bit too much as it's called behind the scenes.